crypto/pkcs7/pkcs7.c

```diff
@@ -843,6 +843,32 @@ int PKCS7_is_detached(PKCS7 *p7) {
   return 0;
 }
 
+int PKCS7_set_detached(PKCS7 *p7, int detach) {
+  GUARD_PTR(p7);
+  if (detach != 0 && detach != 1) {
+    return 0;
+  }
+
+  if (PKCS7_type_is_signed(p7)) {
+    if (p7->d.sign == NULL) {
+      OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_NO_CONTENT);
+      return 0;
+    }
+    if (detach && PKCS7_type_is_data(p7->d.sign->contents)) {
+      ASN1_OCTET_STRING_free(p7->d.sign->contents->d.data);
+      p7->d.sign->contents->d.data = NULL;
+    }
+    return detach;
```

Comment:
The semantic of the `detach` parameter is unexpected. The call always "fails" (returns 0) when `detach = 0`.  

Could we detect `detach == 0` and return 0 at the top? (The difference would be that calls to `OPENSSL_PUT_ERROR` in certain situations would not be made.)
```
if (detach != 1) {
    return 0;
}
```

If the calls to `OPENSSL_PUT_ERROR` are important, we should have test coverage to confirm they are made.

Reply:
The `detach` parameter is expected to be consumed as a boolean. Ruby has a good example of the usage here: https://github.com/ruby/ruby/blob/dd863714bf377b044645ea12b4db48920d49694e/ext/openssl/ossl_pkcs7.c#L497-L500

I completely agree with the weird semantic of the `detach` parameter, especially with `0`. OpenSSL's implementation allows for other numbers to be used, but seeing that it doesn't align with how the API was meant to be used, I decided to restrict the values to only `0` and `1`.
Since the parameter value was meant to take a `boolean` integer, I do think `0` should be allowed as a value though. I'll add coverage for the unsupported operations.